OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs

```diff
@@ -0,0 +1,480 @@
+        this.logger = logger;
+
+        if (!IsInitialized())
+        {
+            try
+            {
+                Initialize();
+            }
+            catch (Exception ex)
+            {
+                // No logging, as this will lead to server crash and log at that point
+                throw new EUSignOAuth2Exception("An error occurred while initializing the cryptographic library.", ex);
+            }
+
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+        }
+
+        if (!IsPrivateKeyLoaded())
+        {
+            try
+            {
+                ReadPrivateKey(
+                    eUSignConfig.PrivateKey.FileName,
+                    eUSignConfig.PrivateKey.MediaType,
+                    eUSignConfig.PrivateKey.MediaDevice,
+                    eUSignConfig.PrivateKey.Password,
+                    eUSignConfig.PrivateKey.JKSAlias,
+                    eUSignConfig.PrivateKey.CertificateFilePaths,
+                    eUSignConfig.PrivateKey.CAIssuerCN);
+            }
+            catch (Exception ex)
+            {
+                // No logging, as this will lead to server crash and log at that point
+                throw new EUSignOAuth2Exception("An error occurred while reading the server's private key.", ex);
+            }
+        }
+    }
+
+    /// <inheritdoc/>
+    public CertificateResponse GetEnvelopCertificateBase64()
+    {
+        if (pkEnvelopCert == null)
+        {
+            logger.LogError("No envelop certificate was loaded");
+            return null;
+        }
+
+        return new()
+        {
+            CertBase64 = Convert.ToBase64String(pkEnvelopCert),
+        };
+    }
+
+    /// <inheritdoc/>
+    public UserInfoResponse DecryptUserInfo(EnvelopedUserInfoResponse encryptedUserInfo)
+    {
+        try
+        {
+            IEUSignCP.CtxDevelopData(
+                pkContext,
+                encryptedUserInfo.EncryptedUserInfo,
+                null,
+                out var developedUserInfo,
+                out _);
+
+            Stream userInfoStream = new MemoryStream(developedUserInfo);
+            return JsonSerializer.Deserialize<UserInfoResponse>(userInfoStream);
+        }
+        catch (Exception ex)
+        {
+            logger.LogError(ex, "An error occurred while decrypting user info");
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Пошук зареєстрованого носія ключової інформації
+    /// This code is from IIT Library usage example with minimal code styling changes.
+    /// </summary>
+    private static void GetKeyMedia(
+        string type,
+        string device,
+        string password,
+        out IEUSignCP.EU_KEY_MEDIA keyMedia)
+    {
+        keyMedia = new IEUSignCP.EU_KEY_MEDIA(0, 0, string.Empty);
+
+        try
+        {
+            int typeIndex, deviceIndex;
+            string curType, curDevice;
+
+            typeIndex = 0;
+            while (true)
```

Comment:
That's very brave idea to not limit `typeIndex` and `deviceIndex` in these `while(true)` cycles

Reply:
You've probably meant that: if device name from the config will be wrong - we will have to wait for some noticeable time while got the very descriptive error.

Reply:
> By default, integral values (such as int, uint, and long) do not throw an exception when they overflow. Instead, they "wrap" -- and this is probably not a value that we want. But we can change this behavior so that an exception is thrown instead. [Link](
https://jeremybytes.blogspot.com/2023/01/checking-for-overflow-in-c.html#:~:text=January%2017%2C%202023-,Checking%20for%20Overflow%20in%20C%23,an%20exception%20is%20thrown%20instead.)

"Noticable" time could last forever.